tests(api): add tests to api gateway

haller33 opened the pull request:
little tests added

gustavosbarreto commented on api/pkg/gateway/context_test.go:
```diff
@@ -0,0 +1,599 @@
+package gateway
+
+import (
+	//	"fmt"
```
Please remove the commented code

gustavosbarreto commented on api/pkg/gateway/context_test.go:
```diff
@@ -0,0 +1,599 @@
+)
+
+func TestRole(t *testing.T){
+
```
Unnecessary new line

gustavosbarreto commented on api/pkg/gateway/context_test.go:
```diff
@@ -0,0 +1,599 @@
+}
+
+func TestTenant(t *testing.T){
+
```
Unnecessary new line

gustavosbarreto commented on api/pkg/gateway/context_test.go:
```diff
@@ -0,0 +1,599 @@
+}
+
+func TestUsername(t *testing.T){
+
```
Unnecessary new line

heiytor commented on api/pkg/gateway/context_test.go:
```diff
@@ -0,0 +1,541 @@
+
+			retStringRole, retBoolRoleValidate := gatewayCtxT.GetRole()
+
+			require.Equal(t, tc.expected.role, retStringRole)
+			require.Equal(t, tc.expected.ok, retBoolRoleValidate)
```
You can compare the entire struct instead of its individual attributes. Change this to:
```go
require.Equal(t, tc.expected, Expected{retStringRole, retBoolRoleValidate})
```

heiytor commented on api/pkg/gateway/gateway_test.go:
```diff
@@ -0,0 +1,214 @@
+
+func TestUsernameFromContext(t *testing.T) {
+	type Expected struct {
+		userNameTenant *models.Username
```
Why `userNameTenant`?

haller33 replied:
I'll be changed, it has been because of an old design decision

heiytor commented on api/pkg/gateway/context_test.go:
```diff
@@ -0,0 +1,541 @@
+			e := echo.New()
+			req := httptest.NewRequest(http.MethodPost, "/", nil)
+
+			req.Header.Set(echo.HeaderContentType, echo.MIMEApplicationJSON)
```
I believe that since you're not sending a body, there's no need to set the `Content-Type` header.

haller33 replied:
Understood

haller33 replied:
and a very good one

heiytor commented on api/pkg/gateway/context_test.go:
```diff
@@ -0,0 +1,541 @@
+				ctxEchoT,
+			}
+
+			retGetTenantFst, retGetTenantSnd := gatewayCtxT.GetTennat()
```
What are `fst` and `snd`?

heiytor commented:
I also noticed that you use `ret` as a prefix for many return values. As far as I know, this is not a common Go pattern, so I believe it would be a good idea to remove it. Additionally, you frequently use `T` as a postfix. Could you explain what this means?

henrybarreto commented on api/pkg/gateway/context_test.go:
```diff
@@ -0,0 +1,541 @@
+				ctxEchoT,
+			}
+
+			retTenantID := gatewayCtxT.Tenant()
```
As mentioned by @heiytor, the `ret` prefix isn't necessary and doesn't follow the Golang pattern for variables in general. I believe it could be simple `tenant` or `tenantID` without cognitive prejudice.

You could check these pattern here: https://go.dev/doc/effective_go#variables

henrybarreto reviewed (commented):
Beyond what already said, I recommend the [Effective Go](https://go.dev/doc/effective_go) to give a better overview of the Golang patterns and recommendations.

haller33 commented on api/pkg/gateway/context_test.go:
```diff
@@ -0,0 +1,541 @@
+
+func TestRole(t *testing.T) {
+	type Expected struct {
+		roleString authorizer.Role
```
sure! understood and agree

haller33 commented:
> I also noticed that you use `ret` as a prefix for many return values. As far as I know, this is not a common Go pattern, so I believe it would be a good idea to remove it. Additionally, you frequently use `T` as a postfix. Could you explain what this means?

this has been a tag name for an abbreviation of 'Tuple' an old intermediate struct created by myself that has been used and later on, removed, and this T has been as a forgotten code, that is now removed

haller33 commented on api/pkg/gateway/context_test.go:
```diff
@@ -0,0 +1,541 @@
+				"X-Role": "owner",
+			},
+			expected: Expected{
+				roleString: authorizer.Role("owner"),
```
Understood

heiytor commented on api/pkg/gateway/context_test.go:
```diff
@@ -0,0 +1,477 @@
+				"X-Role": authorizer.RoleOwner.String(),
+			},
+			expected: authorizer.Role("owner"),
+		}, {
+			description: "validate empty ID string, for ID behaviour",
+			headers: map[string]string{
+				"X-Role": "",
+			},
+			expected: authorizer.RoleInvalid,
+		}, {
+			description: "validate empty headers, for fail function",
+			headers:     map[string]string{},
+			expected:    authorizer.RoleInvalid,
+		},
```
These two cases are redundant. `authorizer.RoleFromString` checks whether `X-Role` corresponds to a valid role. If it doesn't, it returns `authorizer.RoleInvalid`.

Since `echo` assigns an empty string to a non-existent header, both scenarios will produce the same result. If you prefer to maintain distinct cases, you can handle one case without the header (similar to the second one), and another with an invalid role (such as the string `invalid`).

haller33 replied:
Understood

heiytor commented on api/pkg/gateway/context_test.go:
```diff
@@ -0,0 +1,477 @@
+			expected: &models.Tenant{
+				ID: "00000000-0000-4000-0000-000000000000",
+			},
+		}, {
+			description: "validate empty ID string, for ID behaviour",
+			headers: map[string]string{
+				"X-Tenant-ID": "",
+			},
+			expected: nil,
+		}, {
+			description: "validate empty headers, for fail function",
+			headers:     map[string]string{},
+			expected:    nil,
+		},
```
The same applies here. You can safely remove the first case.

Also, you can do the same in the other tests.
